src/borrower/feed/navfeed.sol

```diff
@@ -273,106 +161,272 @@ contract NAVFeed is BaseNFTFeed, Interest, Buckets, FixedPoint {
-        if (buckets[maturityDate_].value == 0 && firstBucket != 0) {
-            _removeBucket(maturityDate_);
+        // return decrease NAV amount
+        return calcDiscount(discountRate.value, safeSub(preFutureValue, fv), uniqueDayTimestamp(block.timestamp), maturityDate_);
+    }
+
+    function borrowEvent(uint loan) public virtual auth {
+        uint risk_ = risk[nftID(loan)];
+
+        // when issued every loan has per default interest rate of risk group 0.
+        // correct interest rate has to be set on first borrow event
+        emit log_named_uint("borrowEvent_loanRates", pile.loanRates(loan));
+        emit log_named_uint("borrowEvent_risk", risk_);
+        if (pile.loanRates(loan) != risk_) {
+            // set loan interest rate to the one of the correct risk group
+            pile.setRate(loan, risk_);
         }
+    }
 
-        // return decrease NAV amount
-        return calcDiscount(safeSub(preFutureValue, fv), uniqueDayTimestamp(block.timestamp), maturityDate_);
+    function unlockEvent(uint loan) public virtual auth {}
+
+    // TODO: we could also remove the writeOffGroupIndex_ argument, loop over the
+    // writeOffGroups and find the latest writeoff group which can be applied.
+    function writeOff(uint loan, uint writeOffGroupIndex_) public {
+        require(!writeOffOverride[loan], "already-overridden");
+
+        bytes32 nftID_ = nftID(loan);
+        WriteOffGroup memory writeOffGroup_ = writeOffGroups[writeOffGroupIndex_];
+        uint maturityDate_ = maturityDate[nftID_];
+        require(block.timestamp >= maturityDate_ + writeOffGroup_.overdueDays * 60 * 60 * 24, "too-early");
+
+        uint currentRate = pile.loanRates(loan);
+        require(currentRate < WRITEOFF_RATE_GROUP_START || writeOffGroup_.overdueDays > writeOffGroups[currentRate].overdueDays, "cannot-go-back");
+
+        _writeOff(loan, writeOffGroupIndex_, nftID_, maturityDate_);
     }
 
-    function calcDiscount(uint amount, uint normalizedBlockTimestamp, uint maturityDate_) public view returns (uint result) {
-        return rdiv(amount, rpow(discountRate.value, safeSub(maturityDate_, normalizedBlockTimestamp), ONE));
+    function overrideWriteOff(uint loan, uint writeOffGroupIndex_) public auth {
+        writeOffOverride[loan] = true;
+        
+        bytes32 nftID_ = nftID(loan);
+        uint maturityDate_ = maturityDate[nftID_];
+        _writeOff(loan, writeOffGroupIndex_, nftID_, maturityDate_);
     }
 
-    function calcTotalDiscount() public view returns(uint) {
-        (uint sum, ) = _calcTotalDiscount();
-        return sum;
+    function _writeOff(uint loan, uint writeOffGroupIndex_, bytes32 nftID_, uint maturityDate_) internal {
+        if (pile.loanRates(loan) < WRITEOFF_RATE_GROUP_START) {
+            // this is the first writeoff, remove future value for loan from bucket
+            uint fv = futureValue[nftID_];
+            buckets[maturityDate_] = safeSub(buckets[maturityDate_], fv);
```

Comment:
This would be only required if the loan is written-off before the maturity date. 

We could safe some gas here.

In that case I think we also need to substract the current PV from the stored totalDiscount